haystack/pipelines/config.py

```diff
@@ -111,9 +127,70 @@ def build_component_dependency_graph(
     return graph
 
 
-def _validate_user_input(input: str):
-    if isinstance(input, str) and not VALID_CODE_GEN_INPUT_REGEX.match(input):
-        raise ValueError(f"'{input}' is not a valid config variable name. Use word characters only.")
+def validate_yaml(path: Path):
+    """
+    Validates the given YAML file using the autogenerated JSON schema.
+
+    :param pipeline_config: the configuration to validate
+    :return: None if validation is successful
+    :raise: `PipelineConfigError` in case of issues.
+    """
+    pipeline_config = read_pipeline_config_from_yaml(path)
+    validate_config(pipeline_config=pipeline_config)
+    logging.debug(f"'{path}' contains valid Haystack pipelines.")
+
+
+def validate_config(pipeline_config: Dict) -> None:
+    """
+    Validates the given configuration using the autogenerated JSON schema.
+
+    :param pipeline_config: the configuration to validate
+    :return: None if validation is successful
+    :raise: `PipelineConfigError` in case of issues.
+    """
+    validate_config_strings(pipeline_config)
+
+    with open(JSON_SCHEMAS_PATH / f"haystack-pipeline-unstable.schema.json", "r") as schema_file:
+        schema = json.load(schema_file)
+
+    compatible_versions = [version["const"].replace('"', "") for version in schema["properties"]["version"]["oneOf"]]
+    try:
+        Draft7Validator(schema).validate(instance=pipeline_config)
+
+        if pipeline_config["version"] == "unstable":
+            logging.warning(
+                "You seem to be using the 'unstable' version of the schema to validate "
+                "your pipeline configuration.\n"
+                "This is NOT RECOMMENDED in production environments, as pipelines "
+                "might manage to load and then misbehave without warnings.\n"
+                f"Please pin your configurations to '{__version__}' to ensure stability."
+            )
+
+        elif pipeline_config["version"] not in compatible_versions:
+            raise PipelineConfigError(
+                f"Cannot load pipeline configuration of version {pipeline_config['version']} "
+                f"in Haystack version {__version__} (compatible versions are {compatible_versions}).\n"
+                "Please check out the documentation "
+                "(https://haystack.deepset.ai/components/pipelines#yaml-file-definitions) "
+                "and fix your configuration accordingly."
+            )
```

Comment:
Actually no, that's the trick :grin: It's not very visible right now because we have a different schema for each version (because the schema actually changed). In practice, if two versions come out with no schema changes, the two versions are going to be compatible JSON-schema-wise and the exception will not trigger. You can see an example in  the `0.7` JSON schema: many versions are supported there. So loading a YAML with version 0.7 in Haystack 1.0 will work just fine.

For DC I wouldn't worry personally, they probably don't care about the warning and/or will stay on non-`rc` releases for production? I'd not recommend them to deploy master continuously anyway :sweat_smile: 

So to summarize, my thinking here was:
- If you're testing out Haystack and don't care about stability, the warning is fine.
- If you're in production, either:
    - use master at your own risk and accept the warning
    - pin the version and review your YAMLs when upgrading.

I think one of the core features of YAML validation is exactly this, to prevent people from upgrading mindlessly and break their pipelines. So I argue this should stay, but happy to talk this through if you prefer.

Reply:
I'm afraid I don't quite understand. If I run:
```python
from haystack.pipelines.config import validate_config

c = {"version": "0.7", "components":[], "pipelines": []}
validate_config(c)
```
it raises 
```
haystack.errors.PipelineConfigError: Validation failed. '0.7' is not valid under any of the given schemas. The error is in 'version'. See the stacktrace for more information.
```
If this is the intended behavior, I'm not really happy with that. Or am I mistaking something?

Reply:
Yep that's the intended behavior. I agree the error is not nice though: I can improve that a lot, something like:

`Version 0.7 is not compatible with haystack 1.2.1rc0. Please upgrade the version in your YAML files and test your pipelines before proceeding (or use 'unstable' to ignore version mismatches)`.

My reasoning is: between 0.7 and 1.2 there has been schema changes, so Haystack 1.2 and Haystack 0.7 will read that file differently. If you believe there aren't any meaningful changes, just change the version and try it out. So:
 - If you don't care about stability, you should be explicit about it and use `unstable`.
 - If you do care about stability, you don't want Haystack 1.2 to load a 0.7 YAML accidentally.

Maybe we should have a chat and hear @julian-risch too. If we're three we can vote :grin:

Reply:
Ah I think I now understand. If we release haystack 1.3 and the schema doesn't change `haystack-pipeline-unstable.schema.json` would be updated to also support `1.3.0`.

But to be honest I am actually not too happy with this behavior. Example:
From version `1.2.0` to `1.2.1rc0` the only changes in schema were:
- Adding WeaviateDocumentStoreComponent
- Adding JoinAnswerComponent
- Adding RouteDocumentsComponent

If I generalize on this, this would mean, that adding a new component results always in a schema change.
Thus adding a new component would always effectively mean that YAMLs of previous versions won't be compatible anymore. Sorry for being so stubborn on this, but is that really what we want here?
I mean, even if you're not using e.g. WeaviateDocumentStore and you have a `1.2.0` YAML you would have to manually update the version to `1.2.1rc0`. Why? 
I'd rather expect that an expection is raised e.g. only if used components unknown to the current haystack version.

Reply:
That's true actually, adding components makes the schema compatible...

I'll see if there's a way to detect only deprecations and use those only as a reason to fail validation. So if Haystack 1.3.0 removes Milvus1, that will count as a schema change, while if it only adds NewDocStore, that would still be backwards compatible. That could solve the issue, right?

There's still the fact that parameters that change behavior won't be detected, which is what worries me more and what even the current validation does not take into account. But that's probably not possible to do with JSON schemas anyway, so I'll ignore this issue.

(PS: I'll weed out older backwards_compatible schemas as I do this modifications. Probably 1.1.0. 1.2.0 and 1.2.1rc0 are all compatible with this logic)

Reply:
Summary:

- Let's break backwards compatibility only if something is removed from the schema (not when something new is added).
- We'll allow unknown components by default in the "static" schemas (the ones saved in `json-schemas`).
- We'll validate custom components by generating a JSON schema for them on the fly and adding it dynamically to `unstable`.

Ofc I'll check the feasibility of all this and report back in case of issues :+1: